src/Access/SettingsConstraints.h

```diff
@@ -87,29 +82,60 @@ class SettingsConstraints
     friend bool operator !=(const SettingsConstraints & left, const SettingsConstraints & right) { return !(left == right); }
 
 private:
-    struct Constraint
+    enum ReactionOnViolation
+    {
+        THROW_ON_VIOLATION,
+        CLAMP_ON_VIOLATION,
+    };
+
+    struct Range
     {
-        std::shared_ptr<const String> setting_name;
-        bool read_only = false;
+        SettingConstraintType type = SettingConstraintType::NONE;
         Field min_value;
         Field max_value;
 
-        bool operator ==(const Constraint & other) const;
-        bool operator !=(const Constraint & other) const { return !(*this == other); }
+        String explain;
+        int code;
+
+        bool operator ==(const Range & other) const;
+        bool operator !=(const Range & other) const { return !(*this == other); }
+
+        bool check(SettingChange & change, const Field & new_value, ReactionOnViolation reaction) const;
+
+        static const Range & allowed()
+        {
+            static Range res;
+            return res;
+        }
+
+        static Range forbidden(const String & explain, int code)
+        {
+            return Range{.type = SettingConstraintType::CONST, .explain = explain, .code = code};
+        }
     };
 
-    enum ReactionOnViolation
+    struct StringHash
     {
-        THROW_ON_VIOLATION,
-        CLAMP_ON_VIOLATION,
+        using is_transparent = void;
+        size_t operator()(std::string_view txt) const
+        {
+            return std::hash<std::string_view>{}(txt);
+        }
+        size_t operator()(const String & txt) const
+        {
+            return std::hash<String>{}(txt);
+        }
     };
+
     bool checkImpl(const Settings & current_settings, SettingChange & change, ReactionOnViolation reaction) const;
 
-    Constraint & getConstraintRef(std::string_view setting_name);
-    const Constraint * tryGetConstraint(std::string_view setting_name) const;
+    Range getRange(const Settings & current_settings, std::string_view setting_name) const;
+
+    // Special container for heterogeneous lookups: to avoid `String` construction during `find(std::string_view)`
+    using RangeMap = std::unordered_map<String, Range, StringHash, std::equal_to<>>;
```

Comment:
Here was some curious logic to make `std::unordered_map<std::string_view>`. I guess it was to avoid allocation while checking rights, I just replaced it with "modern C++". But actually I think it is not so important here to be super fast, so I can completely get rid of it. WDYT?

Reply:
Can't the standard `std::hash` do what's necessary to avoid this allocation?

Reply:
the problem is `std::unordered_map::find(Key)` with `Key=String` is called with `std::string_view`, so conversion is done before even `std::hash`

Reply:
Let's keep your "modern C++" code until we think of a better solution